Docs: update ubuntu example setup for quicker start (#1431)
PowerGridModel/power-grid-model: docs/advanced_documentation/build-guide.md

Reviewer: We may as well do it here as well (maybe there are other instances).
```suggestion
After this section there are examples of setup in Linux (Ubuntu 26.04), Windows 11, and macOS (Tahoe).
```

Reviewer: Can you remove this? It looks really bad and sends the wrong message. This is related to `lcov` and such only, not actual test coverage on MacOS.

Reviewer: ```suggestion

```

To build the library from source, you need to first prepare the compiler toolchains and the build dependencies.
In this section a list of general requirements are given.

> After this section there are examples of setup in Linux (Ubuntu 26.04), Windows 11, and macOS (Tahoe).

Reviewer: ```suggestion
After this section there are examples of setup in Linux (Ubuntu 24.04), Windows 11, and macOS (Tahoe).
```

To list the available presets, run `./build.sh -h`.

## Example Setup for Ubuntu 26.04 (in WSL or physical/virtual machine)

In this section an example is given for setup in Ubuntu 26.04.
You can use this example in Windows Subsystem for Linux (WSL), or in a physical/virtual machine.

While you can use `apt`, setup via [Homebrew](https://brew.sh) has proved to be much simpler.
In addition, the recommended way to get the [C++ packages](#c) and `uv` is via [Homebrew](https://brew.sh).

### Set-up Homebrew

Follow the setup guidelines in [Homebrew](https://brew.sh).
For optimal user experience on Linux, also consider installing the
[requirements](https://docs.brew.sh/Homebrew-on-Linux#requirements) as well.

At the time of writing, the set-up is done as follows:

```shell
sudo apt update && sudo apt -y upgrade                     # make sure you're up to date

> sudo apt-get install build-essential procps curl file git  # install the requirements for Homebrew

Reviewer: After following the `homebrew` installation guide, they only suggest `sudo apt install build-essential` for the requirements.

Also note replacing `apt-get` with the newer `apt`.

Author: ```suggestion
sudo apt install build-essential procps curl file git  # install the requirements for Homebrew
```

Author: > After following the `homebrew` installation guide, they only suggest `sudo apt install build-essential` for the requirements.

Where did you read that?

<img width="1084" height="961" alt="image" src="https://github.com/user-attachments/assets/b168da13-90b5-4e7a-8ccc-3fcaf8746821" />

Reviewer: After running 
```shell
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```
that's what it prompts in my terminal. But if it suggests it like that explicitly, then no problem. Just ammend `apt-get` to `apt`, they should be "equivalent", no?

[requirements](https://docs.brew.sh/Homebrew-on-Linux#requirements) as well.

The following packages are optional depending on your use case:
At the time of writing, the set-up is done as follows:

```shell
sudo apt install -y gcovr lcov # For coverage reports
sudo apt install -y gdb # For debugging
sudo apt install -y wget curl zip unzip tar git # General use tools
sudo apt update && sudo apt -y upgrade                     # make sure you're up to date
sudo apt-get install build-essential procps curl file git  # install the requirements for Homebrew
```

### C++ Dependencies for CMake

The recommended way to get the [C++ packages](#c) and `uv` is via [Homebrew](https://brew.sh):

```shell
brew install boost eigen nlohmann-json msgpack-cxx doctest cmake uv
```
Then install Homebrew using the setup recommendations in [Homebrew](https://brew.sh).

### Environment variables
### C++ Dependencies + Build Requirements

Append the following lines into the file `${HOME}/.bashrc`.
The recommended way to get the [C++ packages](#c) and `uv` is via [Homebrew](https://brew.sh):

```shell
export CXX=clang++-18            # or g++-14

> export CC=clang-18               # or gcc-14

Reviewer: Even if installing `gcc` via `homebrew`, I believe these two environment variables are needed for `scikit-build` as it explicitly expects `CC` and `CXX`, but not totally sure.

Author: right....

export CMAKE_PREFIX_PATH=/home/linuxbrew/.linuxbrew  # only needed for CMake builds
export LLVM_COV=llvm-cov-18      # only if you want to use one of the llvm features

> export CLANG_TIDY=clang-tidy-18  # only if you want to use one of the clang-tidy presets

Author: with the updated brew requirements, this is no longer needed

>     gcc llvm gdb covr lcov lld cmake ninja \

Reviewer: - I don't think we should recommend `gcc` via brew since ` apt install build-essential` already bundles it and can cause confusion. Then the step of adding the environment variables back becomes necessary.
- We had typed `eigen3` instead of `eigen` in many places, can you double check?
- I don't think we should recommend `gdb`, `covr`, and `lcov` by default. If the user needs them, they will know and install independently. Those aren't needed for PGM installation. Let's remove them from our recommendations.
- Do we use `lld` instead of `ld` natively if available? If not, I'd drop it as well.
- `llvm` includes `clang`, right? I think it's good to keep it in, but with the side note that it's only to include `clang` compiler as an alternative.
- Typo: `gcovr` instead of `covr`.

Author: It's an example setup. Users are free to deviate from it (applying to mainly the following 4 bullet points). Ease of setup over minimality

> * I don't think we should recommend `gcc` via brew since ` apt install build-essential` already bundles it and can cause confusion. Then the step of adding the environment variables back becomes necessary.
> * I don't think we should recommend `gdb`, `covr`, and `lcov` by default. If the user needs them, they will know and install independently. Those aren't needed for PGM installation. Let's remove them from our recommendations.
> * Do we use `lld` instead of `ld` natively if available? If not, I'd drop it as well.
> * `llvm` includes `clang`, right? I think it's good to keep it in, but with the side note that it's only to include `clang` compiler as an alternative.

In addition:

> * We had typed `eigen3` instead of `eigen` in many places, can you double check?

The library is called Eigen3, but the package name may differ. Just like how the PGM C API is also named `power_grid_model_c` but the package is not.

> * Typo: `gcovr` instead of `covr`.

```suggestion
    gcc llvm gdb gcovr lcov lld cmake ninja \
```

To list the available presets, run `./build.sh -h`.

> ## Example Setup for Ubuntu 26.04 (in WSL or physical/virtual machine)

Reviewer 2: I would refrain to use Ubuntu 26.04 as an example at this moment. Ubuntu 26.04 is not widely available yet. For example, in WSL, you can only install Ubuntu 26.04 from the WSL listing when 26.04 is ready for point release (usually Sep/Oct of the year). At this moment, you can only manually download a Ubuntu WSL image to install.

Author: It's an example setup - not the "only possible" setup. Ubuntu 26.04 is considered the golden standard for Ubuntu users.

The only reason WSL stays behind is because it's not the default -- although the recommended (note that setting `wsl --set-default-version 2` bumps the default installed version).

As for the GitHub runner images, they are always late. Just the fact that we don't use Ubuntu 26.04 yet in CI doesn't mean we can't recommend our users to use it.

Reviewer 2: No, in WSL it's not about default. You simply cannot run

```powershell
wsl --install Ubuntu-26.04
```

It is not available yet. You have to wait until Sep/Oct of the year for that to be available. That's the design choice made by MS and Carnonical.

Reviewer: I resolved in [6f31db4](https://github.com/PowerGridModel/power-grid-model/pull/1431/commits/6f31db4e99c6ef2b55ae24cb015fd7c084e5e814).

> ## Example Setup for Ubuntu 26.04 (in WSL or physical/virtual machine)

Reviewer: ```suggestion
## Example Setup for Ubuntu 24.04 (in WSL or physical/virtual machine)
```

> In this section an example is given for setup in Ubuntu 26.04.

Reviewer: ```suggestion
In this section an example is given for setup in Ubuntu 24.04.
```

In this section an example is given for setup in Ubuntu 24.04.
In this section an example is given for setup in Ubuntu 26.04.
You can use this example in Windows Subsystem for Linux (WSL), or in a physical/virtual machine.

### Ubuntu Software Packages
While you can use `apt`, setup via [Homebrew](https://brew.sh) has proved to be much simpler.
In addition, the recommended way to get the [C++ packages](#c) and `uv` is via [Homebrew](https://brew.sh).

Install the minimum required packages:
### Set-up Homebrew

```shell
sudo apt update && sudo apt -y upgrade

> sudo apt install -y build-essential gcc g++ clang-18 make ninja-build pkg-config

Reviewer 2: I do think `gcc` and `clang` installation from `apt` is making sense. If you do that, your shared library will be linked to the system `glibc` version, which is important if developers want to build from source and ship prebuilt binary to some other users or environments.

If you install them from `brew`, it will come with its own `glibc` instead of system `glibc`.

Author: still, it's an example setup

in addition, it means that if a user wants to upgrade gcc version to latest (i.e. not shipped by debian), then it's really a hell to set up. the whole goal of this setup scripts is to make basic installs easy for users (including us), not to provide a fully working version that will satisfy all needs of anyone ever compiling PGM.

Reviewer: I resolved in [6f31db4](https://github.com/PowerGridModel/power-grid-model/pull/1431/commits/6f31db4e99c6ef2b55ae24cb015fd7c084e5e814)

sudo apt update && sudo apt -y upgrade                     # make sure you're up to date

> sudo apt install build-essential procps curl file git  # install the requirements for Homebrew

Reviewer: ```suggestion
sudo apt update && sudo apt -y upgrade                     # make sure you're up to date
sudo apt install build-essential gcc g++ clang-18 procps curl file git # install initial dependencies
```

```

Then install Homebrew using the setup recommendations in [Homebrew](https://brew.sh).

### C++ Dependencies + Build Requirements

The recommended way to get the [C++ packages](#c) and `uv` is via [Homebrew](https://brew.sh):

```shell
brew install \

>     gcc llvm gdb gcovr lcov lld cmake ninja \

Reviewer: ```suggestion
    gdb gcovr lcov lld cmake ninja \
```

>     boost eigen nlohmann-json msgpack-cxx doctest cmake uv

Author: ```suggestion
    boost eigen nlohmann-json msgpack-cxx doctest uv
```

```

Append the following lines into the file `${HOME}/.bashrc` (required for `scikit-build` for the Python installation):

```shell
export CXX=clang++  # or g++ or g++-15 or ...

> export CC=clang     # or gcc or gcc-15 or ...

Reviewer: ```suggestion
export CXX=clang++  # or g++ or g++-14 or ...
export CC=clang     # or gcc or gcc-14 or ...
export LLVM_COV=llvm-cov-18      # only if you want to use one of the llvm features
export CLANG_TIDY=clang-tidy-18  # only if you want to use one of the clang-tidy presets
```

In this section an example is given for setup in Ubuntu 24.04.
You can use this example in Windows Subsystem for Linux (WSL), or in a physical/virtual machine.

While you can use `apt`, setup via [Homebrew](https://brew.sh) has proved to be much simpler.
In addition, the recommended way to get the [C++ packages](#c) and `uv` is via [Homebrew](https://brew.sh).

### Set-up Homebrew

Follow the setup guidelines in [Homebrew](https://brew.sh).
For optimal user experience on Linux, also consider installing the
[requirements](https://docs.brew.sh/Homebrew-on-Linux#requirements) as well.

At the time of writing, the set-up is done as follows:

```shell
sudo apt update && sudo apt -y upgrade                     # make sure you're up to date
sudo apt install build-essential gcc g++ clang-18 procps curl file git # install initial dependencies
```

Then install Homebrew using the setup recommendations in [Homebrew](https://brew.sh).

### C++ Dependencies + Build Requirements

The recommended way to get the [C++ packages](#c) and `uv` is via [Homebrew](https://brew.sh):

```shell
brew install \

>     gdb gcovr lcov lld cmake ninja \

Author: i am not sure how it works if `gdb` is installed via `brew` but `gcc` is installed via `apt`. I think it should work but not sure

Reviewer: It should be fine as long as the power-users handles that paths and symlinking correctly. In any case, `gcc` is also available via `homebrew` so I think we can leave this here.

### C++ Dependencies for CMake
Then install Homebrew using the setup recommendations in [Homebrew](https://brew.sh).

### C++ Dependencies + Build Requirements

The recommended way to get the [C++ packages](#c) and `uv` is via [Homebrew](https://brew.sh):

```shell
brew install boost eigen nlohmann-json msgpack-cxx doctest cmake uv
brew install \
    gdb gcovr lcov lld cmake ninja \
    boost eigen nlohmann-json msgpack-cxx doctest uv
```

### Environment variables

Append the following lines into the file `${HOME}/.bashrc`.
Append the following lines into the file `${HOME}/.bashrc` (required for `scikit-build` for the Python installation):

```shell
export CXX=clang++-18            # or g++-14
export CC=clang-18               # or gcc-14

> export CMAKE_PREFIX_PATH=/home/linuxbrew/.linuxbrew  # only needed for CMake builds

Author: do we need to re-add this? people following the homebrew installation proposal will get that by default, but otherwise it is still required

Reviewer: `CC` and `CXX` are needed in spite of the installation method of `clang` (`apt` or `brew`) as pointed out in https://github.com/PowerGridModel/power-grid-model/pull/1431#discussion_r3380721103.

`CMAKE_PREFIX_PATH ` is not needed since we still recommend installing it via brew. If they don't follow and install it via `apt`, then it's on them to fix it. We could add the suggestion but it would probably make the guide more complicated.
